src/library_pthread.js

```diff
@@ -329,9 +325,33 @@ var LibraryPThread = {
       assert(wasmModule instanceof WebAssembly.Module, 'WebAssembly Module should have been loaded by now!');
 #endif
 
+      // When running on a pthread, none of the incoming parameters on the module
+      // object are present. Proxy known handlers back to the main thread if specified.
+      var handlers = [];
+      var knownHandlers = [
+#if expectToReceiveOnModule('onExit')
+        'onExit',
+#endif
+#if expectToReceiveOnModule('onAbort')
+        'onAbort',
+#endif
+#if expectToReceiveOnModule('print')
+        'print',
+#endif
+#if expectToReceiveOnModule('printErr')
+        'printErr',
```

Comment:
So print and printErr were previously not proxied?  What happened when they were called from the thread previously?

Reply:
Indeed, these module handlers were previously not proxied causing the `Module['print']`/`Module['printErr']` handlers to be skipped (i.e. it's always printed on the terminal) on pthreads. See for example the current implementation of `out`/`err`:
```js
var out = Module['print'] || defaultPrint;
var err = Module['printErr'] || defaultPrintErr;
```
https://github.com/emscripten-core/emscripten/blob/30f50a9335267d3bd26120f1eeeab6cf7c908430/src/shell.js#L428-L429

For simple `printf()` usage (without using WasmFS), however, this issue did _not_ occur, since almost all syscalls, including `fd_write()`, are proxied back to the main thread.
https://github.com/emscripten-core/emscripten/blob/30f50a9335267d3bd26120f1eeeab6cf7c908430/src/library_syscall.js#L1087-L1100

This issue occurred with WasmFS, as that implements `stdout`/`stderr` using `_emscripten_out`/`_emscripten_err`, which doesn't have this proxy logic.
https://github.com/emscripten-core/emscripten/blob/30f50a9335267d3bd26120f1eeeab6cf7c908430/system/lib/wasmfs/special_files.cpp#L110-L116
https://github.com/emscripten-core/emscripten/blob/30f50a9335267d3bd26120f1eeeab6cf7c908430/src/library.js#L3364-L3369
(+ it would be a performance penalty if these were always proxied to the main thread, as noticed in https://github.com/emscripten-core/emscripten/issues/17688#issuecomment-1228943702)

Reply:
BTW, this bug only occurs when the handler is set _outside_ the module. For example, it did _not_ occur when emitting the handlers via `--pre-js`, since the pthread worker(s) will `import()` that.

Reply:
Can we avoid the needless proxying in the case when the handler *is* defined in the `--pre-js` maybe?

Reply:
> Can we avoid the needless proxying in the case when the handler _is_ defined in the `--pre-js` maybe?

I'm not sure this can be done without too much hassle. For example, if someone does this in `--pre-js`:

```js
if (!Module.hasOwnProperty('print')) {
  Module['print'] = function(x) {
    console.log(x);
  };
}

if (!Module.hasOwnProperty('printErr')) {
  Module['printErr'] = function(x) {
    console.error(x);
  };
}
```
(i.e. set the default `print`/`printErr` module handlers, if it is not already overridden outside the module - commit 5c2d544c35efd4830c61fe9c9dfcc2e2f2da87be might be relevant here)

It implies that `Module.hasOwnProperty('print')` and `Module.hasOwnProperty('printErr')` both return `true` after that.